qubes-gui-client/src/buffer.rs

```diff
@@ -0,0 +1,389 @@
+            self.state = ReadState::Error;
+            e
+        })
+    }
+
+    pub fn wait(&mut self) {
+        self.vchan.wait()
+    }
+
+    /// If there is nothing to read, return `Ok(None)` immediately; otherwise,
+    /// returns `Ok(Some(msg))` if a complete message has been read, or `Err`
+    /// if something went wrong.
+    pub fn read_header(&mut self) -> io::Result<Option<(Header, &[u8])>> {
+        self.drain()?;
+        let mut ready = self.vchan.data_ready();
+        loop {
+            if ready == 0 {
+                break Ok(None);
+            }
+            match self.state {
+                ReadState::Error => {
+                    break Err(Error::new(ErrorKind::Other, "Already in error state"))
+                }
+                ReadState::ReadingHeader if ready >= size_of::<Header>() => {
+                    let mut header = <Header as Default>::default();
+                    if self.vchan.recv(header.as_mut_bytes()).map_err(|e| {
+                        self.state = ReadState::Error;
+                        e
+                    })? != size_of::<Header>()
+                    {
+                        break Err(Error::new(
+                            ErrorKind::UnexpectedEof,
+                            "Failed to read a full message header",
+                        ));
+                    }
+                    ready -= size_of::<Header>();
+                    let untrusted_len = u32_to_usize(header.untrusted_len);
+                    match qubes_gui::msg_length_limits(header.ty) {
+                        // See below comment regarding empty messages!
+                        None if untrusted_len == 0 => continue,
+                        None => self.state = ReadState::Discard(untrusted_len),
+                        Some(max_len) if max_len.contains(&untrusted_len) => {
+                            // length was sanitized above
+                            self.buffer.resize(untrusted_len, 0);
+                            // If the message has an empty body, **do not wait for a body byte to
+                            // be sent**, as none will ever arrive.  This will cause the code to
+                            // run one message behind, but only for empty messages!
+                            if untrusted_len == 0 {
+                                self.state = ReadState::ReadingHeader;
+                                break Ok(Some((header, &self.buffer[..])));
+                            }
+                            self.state = ReadState::ReadingBody(header, 0)
+                        }
+                        Some(_) => {
+                            break Err(Error::new(
+                                ErrorKind::InvalidData,
+                                "Incoming packet has invalid size",
+                            ))
+                        }
+                    }
+                }
+                ReadState::ReadingHeader => break Ok(None),
```

Comment:
Perhaps handling `ReadState::ReadingHeader if ready < size_of::<Header>()` first would keep the two conditions close enough together to make clear that this branch only applies if not enough data is available? Otherwise this line looks somewhat surprising.

Edit: okay, there's more than one of those in the new branch. Then a comment like `// not enough data` might be enough to clarify...